Inclusion of context object for ES Projection

tobiasschweizer opened the pull request:
I am preparing the inclusion of the composite view's context object. I have some trouble with IntelliJ: The code does not get indexed properly (find usage, click on class def.). @bogdanromanx  Any hints?

I could figure out how to run the relevant tests from within sbt: ` testOnly *CompositeIndexingSpec -- -z "index resources"`

imsdu commented:
Hello :),

Is everything working outside of Intellij ?

Running an `sbt update` and importing the project again in IntelliJ using SBT (not BSP) could fix the problem.

If not, please start a discussion [here](https://github.com/BlueBrain/nexus/discussions) giving your local setup and the result of the Build tab in Intellij after hitting the refresh button in the SBT panel.

tobiasschweizer commented:
<img width="1893" alt="Screenshot 2022-05-09 at 17 21 31" src="https://user-images.githubusercontent.com/6000023/167442715-515c02e5-8231-46d1-ae6d-91a95d640a0e.png">

tobiasschweizer commented:
Great, now it works! I posted the details in https://github.com/BlueBrain/nexus/discussions/3248 in case someone else runs into this ...

tobiasschweizer commented on delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeIndexingStreamEntry.scala:
```diff
@@ -99,21 +103,32 @@ final case class CompositeIndexingStreamEntry(
+          .toCompactedJsonLd(context)
+
+        // TODO: do not exclude context object from document
+        //val ctx = compacted.map(_.ctx.asJson)
```
@bogdanromanx Here, I'd like to combine/merge and return the ctx and obj that were serialised to JSON before. I am not familiar with `circe`, I'd be glad for any hint.

My first approach was to serialise the whole `CompactedJsonLd` to JSON but I think that is not the way to go.

tobiasschweizer replied:
> could you please confirm your intention? As far as I understand/remember, you'd like to retain the @context value in the resulting elasticsearch document, is that correct?

@bogdanromanx Thanks for your reaction.

Yes, that is exactly my intention. When I query the ES projection, I would like to get back the indexed document *including* the JSON-LD context. For performance reasons, we sometimes prefer to do ES queries. However, without the context the results are not self-contained (not directly processable in the RDF sense).

tobiasschweizer replied:
I included the proposed changes in https://github.com/BlueBrain/nexus/pull/3245/commits/6bbfb42bdaedaab98836b84fe62456aa8fb158ca

Please not that I added `asJson` inside `handleCtxInclusion`: `.deepMerge(ctx.contextObj.asJson)`

tobiasschweizer replied:
Now,the spec fails since I set the new `includeContext` to `true`: https://github.com/BlueBrain/nexus/blob/6bbfb42bdaedaab98836b84fe62456aa8fb158ca/delta/plugins/composite-views/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeIndexingSpec.scala#L195

> [info] - should index resources *** FAILED ***
[info]   The code passed to eventually never returned normally. Attempted 37 times over 30.654083506000003 seconds. Last failure message: forAll failed, because: 
[info]     at index 0, Both Json are not equal (ignoring array order)
[info]     {
[info]       "@context" : {
[info]         "@base" : "http://music.com/",
[info]         "@vocab" : "http://music.com/",
[info]         "_constrainedBy" : {
[info]           "@id" : "https://bluebrain.github.io/nexus/vocabulary/constrainedBy",
[info]           "@type" : "@id"
[info]         },
[info]         "_createdAt" : {
[info]           "@id" : "https://bluebrain.github.io/nexus/vocabulary/createdAt",
[info]           "@type" : "http://www.w3.org/2001/XMLSchema#dateTime"
[info]         },
[info]         "_createdBy" : {
[info]           "@id" : "https://bluebrain.github.io/nexus/vocabulary/createdBy",
[info]           "@type" : "@id"
[info]         },
[info]         "_deprecated" : "https://bluebrain.github.io/nexus/vocabulary/deprecated",
[info]         "_incoming" : {
[info]           "@id" : "https://bluebrain.github.io/nexus/vocabulary/incoming",
[info]           "@type" : "@id"
[info]         },
[info]         "_instant" : "https://bluebrain.github.io/nexus/vocabulary/instant",
[info]         "_outgoing" : {
[info]           "@id" : "https://bluebrain.github.io/nexus/vocabulary/outgoing",
[info]           "@type" : "@id"
[info]         },
[info]         "_project" : {
[info]           "@id" : "https://bluebrain.github.io/nexus/vocabulary/project",
[info]           "@type" : "@id"
[info]         },
[info]         "_rev" : "https://bluebrain.github.io/nexus/vocabulary/rev",
[info]         "_schemaProject" : {
[info]           "@id" : "https://bluebrain.github.io/nexus/vocabulary/schemaProject",
[info]           "@type" : "@id"
[info]         },
[info]         "_self" : {
[info]           "@id" : "https://bluebrain.github.io/nexus/vocabulary/self",
[info]           "@type" : "@id"
[info]         },
[info]         "_source" : "https://bluebrain.github.io/nexus/vocabulary/source",
[info]         "_subject" : "https://bluebrain.github.io/nexus/vocabulary/metadata/subject",
[info]         "_updatedAt" : {
[info]           "@id" : "https://bluebrain.github.io/nexus/vocabulary/updatedAt",
[info]           "@type" : "http://www.w3.org/2001/XMLSchema#dateTime"
[info]         },
[info]         "_updatedBy" : {
[info]           "@id" : "https://bluebrain.github.io/nexus/vocabulary/updatedBy",
[info]           "@type" : "@id"
[info]         },
[info]         "_uuid" : "https://bluebrain.github.io/nexus/vocabulary/uuid",
[info]         "by" : {
[info]           "@type" : "@id"
[info]         }
[info]       },
[info]       "@id" : "muse",
[info]       "album" : {
[info]         "@id" : "absolution",
[info]         "title" : "Absolution"
[info]       },
[info]       "genre" : [
[info]         "electronica",
[info]         "progressive rock",
[info]         "alternative rock"
[info]       ],
[info]       "name" : "Muse",
[info]       "start" : 1994
[info]     }
[info]     did not equal
[info]     {
[info]       "@id" : "muse",
[info]       "album" : {
[info]         "@id" : "absolution",
[info]         "title" : "Absolution"
[info]       },
[info]       "genre" : [
[info]         "electronica",
[info]         "progressive rock",
[info]         "alternative rock"
[info]       ],
[info]       "name" : "Muse",
[info]       "start" : 1994
[info]     } (CompositeIndexingSpec.scala:454)

bogdanromanx commented:
Looks pretty good, still requires:
1. adapt existing tests, create new tests that handle the includeContext variation
2. update the api reference to describe the new fields and update examples in the docs to include the new field (https://bluebrainnexus.io/docs/delta/api/views/composite-view-api.html#elasticsearchprojection -> https://github.com/BlueBrain/nexus/blob/master/docs/src/main/paradox/docs/delta/api/views/composite-view-api.md?plain=1)

We can handle the release notes ourselves after merging this PR.

tobiasschweizer commented:
> First-time contributors need a maintainer to approve running workflows. [Learn more.](https://docs.github.com/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks) 1 expected check

@bogdanromanx @imsdu Could you approve the workflows so the test I added gets executed on GitHub? Thx.!

tobiasschweizer commented:
@bogdanromanx @imsdu I can see that the checks are running now :-)

bogdanromanx commented:
Yep, just approved the workflow. :)

tobiasschweizer commented:
reformatting done in 4f6eeb85c64c400eb466546ffb80d22a37dbbed8

bogdanromanx commented:
You can format the main code with:
```
sbt scalafmt
```
and test code with:
```
sbt Test/scalafmt
```

tobiasschweizer commented:
@bogdanromanx I added a new view id for the test I added, see e7afaf5159e55400489041f8b611a9b4bd039d40

I am not sure about the dependencies between the existing test, i.e. a view is created, updated, and then maybe read. So reusing the existing id for an update may break this.

tobiasschweizer commented:
@bogdanromanx with `sbt compositeViewsPlugin/test`, I have three failing tests:

> [info] *** 3 TESTS FAILED ***
[error] Failed tests:
[error] 	ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.serialization.EventSerializerSpec
[error] 	ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.routes.CompositeViewsRoutesSpec
[error] 	ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.CompositeViewsSpec
[error] (compositeViewsPlugin / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 68 s (01:08), completed 22 Jul 2022, 08:41:53

bogdanromanx commented:
The current CI failures for docs and integration tests are unrelated to the changes here.
